File: day3.scala
//> using file ./parsers.scala

package io.cjf.aoc2024.day3
import io.cjf.aoc2024.parsers.{Parser, ParseResult}
import scala.io.Source

val inputs = Source.fromFile("./inputs/day3.txt").mkString

val Mul = (Parser.lit("mul(") >> Parser.num << Parser.lit(
  ","
)) :+ (Parser.num << Parser.lit(")"))

def part1(inp: String): Int =
  Mul(inp) match
    case ParseResult.Success((a, b), rest) =>
      a * b + (if rest.length > 0 then part1(rest) else 0)
    case ParseResult.Failure =>
      if inp.length > 0 then part1(inp.slice(1, inp.length)) else 0

enum ParseState:
  case Enabled
  case Disabled

def part2(inp: String, state: ParseState): Int =
  if inp.isEmpty then 0
  else
    state match
      case ParseState.Enabled =>
        (Parser.lit("don't()") | Mul)(inp) match
          case ParseResult.Failure => part2(inp.slice(1, inp.length), state)
          case ParseResult.Success(v, rest) =>
            v match
              case "don't()" => part2(rest, ParseState.Disabled)
              case (a, b)    => a * b + (part2(rest, state))
              case _         => throw new Error("Unreachable")
      case ParseState.Disabled =>
        Parser.lit("do()")(inp) match
          case ParseResult.Success(v, rest) => part2(rest, ParseState.Enabled)
          case ParseResult.Failure => part2(inp.slice(1, inp.length), state)

@main def main() =
  println(s"Part 1: ${part1(inputs)}")
  println(s"Part 2: ${part2(inputs, ParseState.Enabled)}")

File: src/bin/day3.rs
use adventofcode2024::parsers::{Either, ParseResult, Parser, Parsers};

fn inputs() -> String {
    std::fs::read_to_string("./inputs/day3.txt").unwrap()
}

fn mul() -> impl Parser<Target = (u64, u64)> {
    Parsers::lit("mul(")
        .then(Parsers::num())
        .followed_by(",")
        .and(Parsers::num().followed_by(")"))
}

fn part1(inp: &str) -> u64 {
    match mul().apply(inp) {
        ParseResult::Success {
            value: (a, b),
            rest,
        } => a * b + (if rest.is_empty() { 0 } else { part1(rest) }),
        ParseResult::Failure => {
            if !inp.is_empty() {
                part1(&inp[1..])
            } else {
                0
            }
        }
    }
}

#[derive(Clone, Copy)]
enum ParseState {
    Enabled,
    Disabled,
}

fn part2(inp: &str, state: ParseState) -> u64 {
    if inp.is_empty() {
        return 0;
    }
    match state {
        ParseState::Disabled => match Parsers::lit("do()").apply(inp) {
            ParseResult::Failure => part2(&inp[1..], state),
            ParseResult::Success { rest, .. } => part2(rest, ParseState::Enabled),
        },
        ParseState::Enabled => match Parsers::lit("don't()").or(mul()).apply(inp) {
            ParseResult::Failure => part2(&inp[1..], state),
            ParseResult::Success { value, rest } => match value {
                Either::Left(..) => part2(rest, ParseState::Disabled),
                Either::Right((a, b)) => a * b + part2(rest, state),
            },
        },
    }
}

fn main() {
    println!("Part 1: {}", part1(&inputs()));
    println!("Part 2: {}", part2(&inputs(), ParseState::Enabled));
}
